metricbeat/module/system/cpu/metrics/metrics.go

```diff
@@ -0,0 +1,249 @@
+
+	// Subtract wait time from total
+	// Wait time is not counted from the total as per #7627.
+	if metrics.currentSample.wait != nil {
+		idleTime = idleTime + cpuMetricTimeDelta(metrics.previousSample.wait, metrics.currentSample.wait, timeDelta, numCPU)
+	}
+
+	totalPct := common.Round(float64(numCPU)-idleTime, common.DefaultDecimalPlacesCount)
+
+	event.Put("total"+pathPostfix, totalPct)
+
+	fn := func(field reflect.Value, iter int, name string) {
+		mapName := fmt.Sprintf("%s%s", name, pathPostfix)
+		valueOfPrevCPU := reflect.ValueOf(metrics.previousSample)
+		prevValue := valueOfPrevCPU.Field(iter)
+		var prevUint uint64
+		if !prevValue.IsNil() {
+			prevUint = prevValue.Elem().Uint()
+		}
+		currentValue := field.Uint()
+		event.Put(mapName, cpuMetricTimeDelta(&prevUint, &currentValue, timeDelta, numCPU))
+	}
+
+	metrics.currentSample.iterateCPUWithFunc(fn)
+}
+
+/*
+	Ticks(), Percentages(), and NormalizedPercentages()
+	are wrappers around OS-specific implementations that are meant to insure
+	we only return data that is appropriate for a given OS.
+
+	To implement this API for a new OS, you must supply a Get(string) (CpuMetrics, error) function
+	This function should return a CPUMetrics function where any unavailable metric is nil
+*/
+
+// NormalizedPercentages fills a given MapStr with normalized CPU usage percentages
+func (m *Metrics) NormalizedPercentages(event *common.MapStr) {
+	// "normalized" in this sense means when we multiply/subtract by the CPU count, we're getting percentages that amount to the average usage per-cpu, as opposed to system-wide
+	normCPU := 1
+
+	timeDelta := m.currentSample.Total() - m.previousSample.Total()
+	if timeDelta <= 0 {
+		return
+	}
+
+	m.fillCPUMetrics(event, normCPU, timeDelta, ".norm.pct")
+}
+
+// Percentages fills a given MapStr with CPU usage percentages
+func (m *Metrics) Percentages(event *common.MapStr) {
+	timeDelta := m.currentSample.Total() - m.previousSample.Total()
+	if timeDelta <= 0 {
+		return
+	}
+	// on per-core metrics, is doesn't make sense to have global counts, since the data itself is per-core
+	// if this is called from a metric we got with FetchCores(), normalize it.
+	normCPU := m.count
+	if !m.isTotals {
+		normCPU = 1
+	}
+	m.fillCPUMetrics(event, normCPU, timeDelta, ".pct")
+}
+
+// Ticks fills a given MapStr with CPU tick counts
+// This value is calculated on windows, and comes directly from OS APIs on other platforms
+func (m *Metrics) Ticks(event *common.MapStr) {
+	m.currentSample.fillTicks(event)
+}
+
+// CPUCount returns the count of CPUs. When available, use this instead of runtime.NumCPU()
+func (m *Metrics) CPUCount() int {
+	return m.count
+}
+
+// cpuMetricTimeDelta is a helper used by fillTicks to calculate the delta between two CPU tick values
+func cpuMetricTimeDelta(v0, v1 *uint64, timeDelta uint64, numCPU int) float64 {
+	var prev, current uint64
+	if v0 == nil {
+		prev = 0
```

Comment:
On what cases `v0` and `v1` can be nil? If this can happen in the first collection, would setting `prev` to 0 produce peaks?

Reply:
Would this be different in current implementation?

Reply:
>  If this can happen in the first collection, would setting prev to 0 produce peaks?

This is more or less what happens in the current code. Everything is `0` in `prev` on first collection, and all the percentages are `0` as a result.

Reply:
Ok, if only percentages are 0 in the first collection I am ok with that. I was wondering if also some reported ticks would be huge (as of `current - 0` for a given high `current`). If we have problematic first values maybe we can skip the first collection.

Reply:
> If we have problematic first values maybe we can skip the first collection.

Huh, never considered that. We might want to do something like that, although I've never seen any weirdness in the wild.

Reply:
Ok, if we are not generating weird values it LGTM.